File: 2021/7/solutions/python/slate/d7.py
import sys

input = sys.stdin.readlines()
inp = input[0].split(',')

median = []

for i in inp:
    median.append(int(i))

median = sorted(median)
mid = int(round(len(median) // 2, 1))


fuel = 0

gold = int(median[mid])

for i in median:
    if int(i) > gold:
        dist = int(i) - gold
        fuel += dist
    elif gold > int(i):
        dist = gold - int(i)
        fuel += dist


print(fuel)

gold = 0
gold_counter = 0
for med in median:
    gold += med
    gold_counter += 1

gold = round((gold * 999 / gold_counter) / 1000)

fuel = 0
for i in median:
    if int(i) > gold:
        dist = int(i) - gold
        for j in range(dist + 1):
            fuel += j
    elif gold > int(i):
        dist = gold - int(i)
        for j in range(dist + 1):
            fuel += j

print(fuel)

File: 2021/7/solutions/golang/slate/d7/d7.go
package d7

import (
	"log"
	"math"
	"sort"
	"strconv"
	"strings"
)

type D7 struct {
}

func (d D7) PartOne(in []string) (int, int) {
	input := strings.Split(in[0], ",")

	var median []int

	for _, i := range input {
		medianint, err := strconv.Atoi(i)
		if err != nil {
			log.Fatal(err)
		}
		median = append(median, medianint)
	}

	sort.Ints(median)
	mid := int(len(median) / 2)

	fuel := 0

	gold := median[mid]

	for _, i := range median {
		if i > gold {
			dist := i - gold
			fuel += dist
		} else if gold > i {
			dist := gold - i
			fuel += dist
		}
	}

	gold = 0
	gold_counter := 0

	for _, med := range median {
		gold += med
		gold_counter += 1
	}

	gold = int(math.Round(float64(((gold * 999) / gold_counter)) / 1000))

	fuel2 := 0
	for _, i := range median {
		if i > gold {
			dist := i - gold
			for j := 0; j <= dist; j++ {
				fuel2 += j
			}
		} else if gold > i {
			dist := gold - i
			for j := 0; j <= dist; j++ {
				fuel2 += j
			}
		}
	}
	return fuel, fuel2
}
